src/test/java/org/checkstyle/autofix/RemoveViolationComments.java

```diff
@@ -49,21 +48,31 @@ public TreeVisitor<?, ExecutionContext> getVisitor() {
     private static final class ViolationCommentRemover extends JavaIsoVisitor<ExecutionContext> {
         @Override
         public Space visitSpace(Space space, Space.Location loc, ExecutionContext ctx) {
-            final StringBuilder suffixAccumulator = new StringBuilder();
+            String suffixAccumulator = null;
+            final List<Comment> filteredComments = new ArrayList<>();
 
-            final List<Comment> filteredComments = space.getComments().stream()
-                    .map(comment -> {
-                        Comment result = comment;
-                        if (!comment.isMultiline() && comment instanceof TextComment) {
-                            final TextComment textComment = (TextComment) comment;
-                            if (textComment.getText().startsWith("violation")) {
-                                suffixAccumulator.append(textComment.getSuffix());
-                                result = null;
-                            }
+            for (Comment comment : space.getComments()) {
+                Comment result = comment;
+                if (!comment.isMultiline() && comment instanceof TextComment) {
+                    final TextComment textComment = (TextComment) comment;
+                    if (textComment.getText().matches("\\s*(\\d+\\s*)?violation.*")) {
+                        if (!filteredComments.isEmpty()) {
+                            final int lastIndex = filteredComments.size() - 1;
+                            final Comment previousComment = filteredComments.get(lastIndex);
+                            filteredComments.set(lastIndex,
+                                    previousComment.withSuffix(textComment.getSuffix()));
                         }
-                        return result; })
-                    .filter(Objects::nonNull)
-                    .collect(Collectors.toList());
+                        else {
+                            suffixAccumulator = textComment.getSuffix();
+                        }
+                        result = null;
```

Comment:
instead of setting it null and later checking for null you may extend the outer if condition with `else { filteredComments.add(result); }`